impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.impl.persistence;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import org.ehcache.CachePersistenceException;
+import org.ehcache.impl.internal.concurrent.ConcurrentHashMap;
+import org.ehcache.spi.persistence.StateRepository;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.FilenameFilter;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * FileBasedStateRepository
+ */
+class FileBasedStateRepository implements StateRepository, Closeable {
+
+  public static final String MAP_FILE_SUFFIX = "-map.bin";
+  private final File dataDirectory;
+  private final ConcurrentMap<String, ConcurrentMap<?, ?>> knownMaps;
+
+  FileBasedStateRepository(File directory) throws CachePersistenceException {
+    if (directory == null) {
+      throw new NullPointerException("directory must be non null");
+    }
+    if (!directory.isDirectory()) {
+      throw new IllegalArgumentException(directory + " is not a directory");
+    }
+    this.dataDirectory = directory;
+    knownMaps = new ConcurrentHashMap<String, ConcurrentMap<?, ?>>();
+    loadMaps();
+  }
+
+  @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
+  private void loadMaps() throws CachePersistenceException {
+    try {
+      for (File file : dataDirectory.listFiles(new FilenameFilter() {
+        @Override
+        public boolean accept(File dir, String name) {
+          return name.endsWith(MAP_FILE_SUFFIX);
+        }
+      })) {
+        FileInputStream fis = new FileInputStream(file);
+        try {
+          ObjectInputStream oin = new ObjectInputStream(fis);
+          try {
+            ConcurrentMap<?, ?> map = (ConcurrentMap<?, ?>) oin.readObject();
+            knownMaps.put(file.getName().replace(MAP_FILE_SUFFIX, ""), map);
```

Comment:
Non case preserving filesystems will bite you here.  The map name will need to be stored in the file itself instead, and then name will need something like an md5 to avoid clashes due to casing.